Keyd application mapping

Doosty opened the pull request:
## Description of changes
Properly handle the keyd-application-mapper python script and add settings for it.

## Things done

- wrap python script so that it can access `keyd` by name
- add users to group `keyd`, this is mandatory so that the script running in the userspace is allowed to bind
- rename socket to `keyd.socket`, the keyd.sock doesnt exist anymore i guess
- add systemd user service for `keyd-application-mapper`, make sure to keep its environment PATH
- migrate the `keyd-application-mapper` configs from .config to etc, so we dont need home-manager to run it
- add service that changes permissions of `/etc/keyd` so that the script can access them if user is in group `keyd`
- add module settings for the application mapping
- removed the systemd hardening because it somehow didnt let the process set the guid, someone should fix that and readd it

- Built on platform(s)
  - [ x ] x86_64-linux
  - works on my machine (wayland, hyprland)
 
 ### Needs testing
Other maintainers should thoroughly test this and probably make changes, but just throwing this out there because i had it locally for too long and i am irked that the one in nixpkgs doesnt support keyd-application-mapper. This is my first commit please tell me if i did something wrong. Cheers.
@peterhoeg , @JohnAZoidberg , @woojiq

Doosty committed:
Update default.nix

wrap the script so it has access to calling keyd, migrate its config files to etc

Doosty committed:
Update keyd.nix

fix keyd communication with its script (group and socket), add module settings and user services for keyd-application-mapper

woojiq commented:
Good job. I'll be able to test it later (maybe next weekend), it looks good from a quick review.
But you removed all systemd hardening, which is not good. I guess you can only remove `RestrictSUIDSGID`. See https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html#RestrictSUIDSGID=

Doosty committed:
Update default.nix

change the substituting to a shorter version

Doosty committed:
Update keyd.nix

re-add the systemd hardening but with the thing that was causing problems commented out, add a couple of assertions, some minor changes to the module and its descriptions

woojiq commented on pkgs/tools/inputmethods/keyd/default.nix:
```diff
@@ -57,13 +62,15 @@ stdenv.mkDerivation {
+
   enableParallelBuilding = true;
 
   # post-2.4.2 may need this to unbreak the test
   # makeFlags = [ "SOCKET_PATH/run/keyd/keyd.socket" ];
```
outdated comment

woojiq commented on nixos/modules/services/hardware/keyd.nix:
```diff
@@ -126,10 +174,10 @@ in
       # post-2.4.2 may need to set makeFlags in the derivation:
       #
       #     makeFlags = [ "SOCKET_PATH/run/keyd/keyd.socket" ];
```
outdated comment

woojiq commented on nixos/modules/services/hardware/keyd.nix:
```diff
@@ -140,10 +188,8 @@ in
-
         # Hardening
-        CapabilityBoundingSet = [ "CAP_SYS_NICE" ];
+        # CapabilityBoundingSet = [ "CAP_SYS_NICE" ];
```
This is not a complete list of capabilities, so just remove one if you don't need it :)

woojiq reviewed (commented):
I don't know how to test this without rebuilding my entire system (probably vm will help), and I can't do this right now. In general I think it would be a good idea to write a simple test (as I wrote for keyd) that just checks if the service is running + simple mapping. This way we are protected from regressions and it is easy to know if a new update of keyd package works. So I advise you to write something like this for application mapper. Otherwise, you need to check manually every time you change some parameter.
